### PR TITLE
Rework CMake enviroment

### LINKED ISSUE #2619: The current CMake enviroment is heavily flawed
Most people in the SuperTux team might know me as a "possible cmake expert". And I say that the current way dependencies are managed is **terrible!**

The only platform that is easy to build for is Linux, because of the abundance of a package manager and pkgconfig. Building SuperTux for Windows is a massive pain. You either:
- Use vcpkg in hopes of it working
- Build every possible dependency from source, one after another

Both of these options still require you to modify the CMake scripts, for some weird reason, and to do so, one needs to have significant knowdlege about CMake.

My proposal would be to build as many dependencies from source as possible using git submodules and `add_subdirectory`, regardless of platform. Therefore it's guaranteed to work flawlessly (if managed correctly) on all platforms (the only platform I have doubts with is webassembly). It would also help my [new cmake-based android builds](https://github.com/SuperTux/supertux/pull/2485).

Sure, building dependencies from source might be slow, but with the Ninja build system, you don't even notice you're building dependencies from source.

I might come out with a pull request taking a look at all of this, but I would like to know your opinion.

### PR DESCRIPTION
**Pull request by MatusGuy**
Fixes #2619
- [ ] Build systems and generators
  - [x] Ninja
  - [ ] GNU Make
  - [ ] MinGW Make
  - [ ] NMake/JOM
- [ ] Compiling & running
  - [ ] Linux host
    - [x] Clang
    - [ ] GCC
    - [ ] MinGW
  - [ ] Windows host
    - [x] clang-cl
    - [ ] ~~Clang~~
    - [ ] ~~LLVM MinGW~~
    - [x] MSVC
    - [ ] MinGW
  - [ ] MacOS host
- [ ] Packaging
  - [ ] Windows
    - [ ] MSI
    - [ ] Portable ZIP
  - [ ] Linux
    - [ ] AppImage
  - [ ] MacOS
- [ ] Fix workflows
- [ ] FetchContent
- [ ] Better & simpler git detection

### CODE REVIEW

**Comment by tobbi** [comment]
...

**Comment by MatusGuy** [comment]
@tobbi sorry if it looks weird im just getting things ready

**Comment by tobbi** [comment]
I think you're doing this to troll people. 

But the more I look at it, the more I'm convinced that *this* is "the pull request" I've mentioned for the longest time. Congratulations. I may have to reverse my decisions about this being the pull request.

**Comment by tylerandari13** [comment]
> But the more I look at it, the more I'm convinced that _this_ is "the pull request" I've mentioned for the longest time.

If this is the pull request, then whats "the issue"?

**Comment by MatusGuy** [comment]
@tobbi ill prove it to you

**Comment by MatusGuy** [comment]
i truly shouldn'tve deleted those files like that so sorry to anybody i scared

**Review by tobbi — DISMISSED** [review]
Go back in time and undelete those files.

**Comment by MatusGuy** [comment]
> Go back in time and undelete those files.

they arent gonna be used but ok

**Comment by tobbi** [comment]
Maybe I should mark my jokes a little better

**Comment by MatusGuy** [comment]
tobbi may i clarify that you sounded even grumpier than usual (idk how this is even possible) when you called this a troll

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+set(BUILD_CONFIG_DATA_DIR "${PROJECT_BINARY_DIR}/data")
+
+## Add definitions
+if(CMAKE_BUILD_TYPE MATCHES Rel)
+  option(IS_SUPERTUX_RELEASE "Build as official SuperTux release" ON)
+  add_definitions(-DRELEASE)
+else()
+  option(IS_SUPERTUX_RELEASE "Build as official SuperTux release" OFF)
+  add_definitions(-DDEBUG)
+endif()
```
Comment (mrkubax10) [inline CMakeLists.txt:33]: Maybe instead of duplicating `option`, use helper variable like `IS_SUPERTUX_RELEASE_DEFAULT` and then pass that to `option`.

**Comment by MatusGuy** [comment]
FCK

**Comment by MatusGuy** [comment]
ONE DAY! GRUMBEL WILL BE FANUM TAXED DOWN THE SKIBIDI TOILET!!!

**Comment by Frostwithasideofsalt** [comment]
Marty are you okay

**Comment by MatusGuy** [comment]
yEA

**Comment by tylerandari13** [comment]
Despite all odds, theres still 1 successful commit.

**Comment by MatusGuy** [comment]
Do y'all think it'd be a good idea to use FetchContent instead of git submodules?

**Comment by Vankata453** [comment]
> Do y'all think it'd be a good idea to use FetchContent instead of git submodules?

I'd prefer submodules, because with them I can edit the source code of dependencies locally and compile with the game right away.

**Comment by MatusGuy** [comment]
... and you can do that fetchcontent too. All fetchcontent does is download a zip and uncompress it. It's better than submodules because you're not cloning the history of the repository, you're just downloading a zip.

**Comment by tobbi** [comment]
> Do y'all think it'd be a good idea to use FetchContent instead of git submodules?

I don't know but here's a German reddit thread about the worst drug experiences:
https://old.reddit.com/r/FragReddit/comments/1arfobr/was_war_eure_schlimmste_erfahrung_mit_drogen/

**Comment by MatusGuy** [comment]
Tobbi what the fuck

**Comment by Vankata453** [comment]
> ... and you can do that fetchcontent too. All fetchcontent does is download a zip and uncompress it. It's better than submodules because you're not cloning the history of the repository, you're just downloading a zip.

Then I would be fine with it. However, perhaps managing the commit, pointed to, is easier with submodules? Otherwise, you probably have to edit it from the CMake file. But it likely wouldn't be an issue.

**Comment by MatusGuy** [comment]
Do you want to push the changes in the submodules?

**Comment by Vankata453** [comment]
Actually, yeah. I definitely see it as more convenient.

**Comment by MatusGuy** [comment]
fetchcontent or pushing from submodule?

**Comment by Vankata453** [comment]
Pushing from submodule, as you mentioned. I can recall doing that a lot.

**Comment by tobbi** [comment]
Through this PR I learned that you can report code for crypto mining. I just did.

**Comment by MatusGuy** [comment]
tobbi this pr will fix all macos ci failures

**Comment by tobbi** [comment]
> tobbi this pr will fix all macos ci failures

Jokes on you: Mac OS CI failure were fixed yesterday bi Xi Xing Ping.

**Comment by MatusGuy** [comment]
I'll break them again

**Comment by tylerandari13** [comment]
Sometimes you need to tear something down so you can rebuild it from scratch even better.

**Comment by tobbi** [comment]
> Sometimes you need to tear something down so you can rebuild it from scratch even better.

I do that a lot with other people. But I won't rebuild them.

**Comment by tobbi** [comment]
Also remember: There is no market. It's all priced in.

**Comment by tobbi** [comment]
Can you fix conflicts so I can trigger builds?

**Comment by MatusGuy** [comment]
100% sure the builds will just fail. I can't fix conflicts rn.

**Comment by MatusGuy** [comment]
@tobbi if you want you can try cloning and compiling this on your machine. I would appreciate it. I don't recommend using a branch.

**Comment by tobbi** [comment]
Why not?

**Comment by MatusGuy** [comment]
it adds new submodules and if you use branches the submodules will start getting real funky.

**Comment by tobbi** [comment]
What do you mean "real funky"? It's commit references, how bad can it get?

**Comment by MatusGuy** [comment]
Idk. I just have a tedious experience with one repository. You can do whatever u want

**Comment by MatusGuy** [comment]
I'm not supporting clang for windows:
- llvm-rc is not compatible with all common rc compiler features
- some cmake files do not address clang for windows

**Comment by MatusGuy** [comment]
@Vankata453 do you still want submodules? Or can we switch to fetchcontent once you're done with simplesquirrel?

**Comment by Vankata453** [comment]
> @Vankata453 do you still want submodules? Or can we switch to fetchcontent once you're done with simplesquirrel?

I find submodules convenient, wish we could still use them to be honest.

**Comment by MatusGuy** [comment]
i still don't get for what reason though. They waste a lot of space

**Comment by Vankata453** [comment]
> Pushing from submodule, as you mentioned. I can recall doing that a lot.

This, but if they're taking up too much space...

**Comment by MatusGuy** [comment]
can we only make a submodule for simplesquirrel then

**Comment by Vankata453** [comment]
> can we only make a submodule for simplesquirrel then

It's not just about simplesquirrel, it's about being able to edit any component of the game locally. Unless you can make it so it supports fetchcontent by default and submodules, only if they exist.

**Comment by Vankata453** [comment]
That would mean we have to update both fetchcontent and submodule target commits, but it shouldn't be a problem, I believe. A flag like `-DUSE_SUBMODULES` which is `OFF` by default would allow switching.

**Comment by MatusGuy** [comment]
i think you're making this more complicated than it needs to be. i can't see why you would need to push commits anywhere else than simplesquirrel.

**Comment by Vankata453** [comment]
Would it be complicated to check whether `external/submodule/CMakeLists.txt` exists and if not, use fetchcontent? And that option I mentioned earlier wouldn't even be needed, you can just not clone them if you don't want to.

**Comment by MatusGuy** [comment]
i feel like you'd be adding bloat by using submodules like that

**Comment by Vankata453** [comment]
What bloat?

**Comment by MatusGuy** [comment]
nvm. sure we can use "both". ill see if i like this in the end

**Comment by MatusGuy** [comment]
I'm also not supporting LLVM MinGW because its xinput file conflicts with sdl's SDL_xinput.h and i could not find a way to disable it.

**Pull request closed without merging**